runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateCache.java

```diff
@@ -116,222 +129,198 @@ private ForComputation(String computation) {
+      localCache = new HashMap<>();
     }
-    if (entry.getCacheToken() != cacheToken) {
-      stateCache.invalidate(id);
-      return null;
+
+    public String getStateFamily() {
+      return stateFamily;
     }
-    if (workToken <= entry.getLastWorkToken()) {
-      // We don't used the cached item but we don't invalidate it.
-      return null;
+
+    public <T extends State> @Nullable T get(StateNamespace namespace, StateTag<T> address) {
+      StateId id = new StateId(forKey, stateFamily, namespace);
+      @SuppressWarnings("nullness") // Unsure how to annotate lambda return allowing null.
+      StateCacheEntry entry = localCache.computeIfAbsent(id, key -> stateCache.getIfPresent(key));
+      return entry == null ? null : entry.get(namespace, address);
     }
-    return entry.get(namespace, address);
-  }
 
-  private <T extends State> void put(
-      WindmillComputationKey computationKey,
-      String stateFamily,
-      long cacheToken,
-      long workToken,
-      StateNamespace namespace,
-      StateTag<T> address,
-      T value,
-      long weight) {
-    StateId id = new StateId(computationKey, stateFamily, namespace);
-    StateCacheEntry entry = stateCache.getIfPresent(id);
-    if (entry == null) {
-      synchronized (this) {
-        keyIndex.put(id.getWindmillComputationKey(), id);
+    public <T extends State> void put(
+        StateNamespace namespace, StateTag<T> address, T value, long weight) {
+      StateId id = new StateId(forKey, stateFamily, namespace);
+      @Nullable StateCacheEntry entry = localCache.get(id);
+      if (entry == null) {
+        entry = stateCache.getIfPresent(id);
+        if (entry == null) {
+          entry = new StateCacheEntry();
+        }
+        boolean hadValue = localCache.putIfAbsent(id, entry) != null;
+        assert (!hadValue);
       }
+      entry.put(namespace, address, value, weight);
     }
-    if (entry == null || entry.getCacheToken() != cacheToken) {
-      entry = new StateCacheEntry(cacheToken);
-      this.displayedWeight += id.getWeight();
-      this.displayedWeight += entry.getWeight();
+
+    public void persist() {
+      localCache.forEach((id, entry) -> stateCache.put(id, entry));
     }
-    entry.setLastWorkToken(workToken);
-    this.displayedWeight += entry.put(namespace, address, value, weight);
-    // Always add back to the cache to update the weight.
-    stateCache.put(id, entry);
   }
 
-  /** Struct identifying a cache entry that contains all data for a key and namespace. */
-  private static class StateId implements Weighted {
+  /** Returns a per-computation view of the state cache. */
+  public ForComputation forComputation(String computation) {
+    return new ForComputation(computation);
+  }
 
-    private final WindmillComputationKey computationKey;
+  /**
+   * Struct identifying a cache entry that contains all data for a ForKey instance and namespace.
+   */
+  private static class StateId implements Weighted {
```

Comment:
Wondering whether it's worth using AutoValue for some of these classes. With autovalue you can simply annotate the hashCode method with @Memoize to avoid recomputing it.

Reply:
Just leaving as is for now.
Since we always need the hash, might as well calculate up front and use for equals etc too instead of just memoizing